pkg/util/validation/exporter/ring_test.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package exporter
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/kv/consul"
+	"github.com/grafana/dskit/ring"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestOverridesExporter_emptyRing(t *testing.T) {
+	ringStore, closer := consul.NewInMemoryClient(ring.GetCodec(), log.NewNopLogger(), nil)
+	t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+	// Create an empty ring.
+	ctx := context.Background()
+	require.NoError(t, ringStore.CAS(ctx, ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		return ring.NewDesc(), true, nil
+	}))
+
+	cfg := RingConfig{Enabled: true}
+	cfg.KVStore.Mock = ringStore
+
+	cfg.InstanceID = "instance-1"
+	cfg.InstanceAddr = "127.0.0.1"
+	i1, err := newRing(cfg, log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, i1.client))
+	t.Cleanup(func() { require.NoError(t, services.StopAndAwaitTerminated(ctx, i1.client)) })
+
+	_, err = i1.isLeader()
+	require.ErrorIs(t, err, ring.ErrEmptyRing)
+}
+
+// TestOverridesExporterRing_scaleDownAndUp tests that a maximum of one leader
+// replica exists at any point in time while the number of replicas is scaled.
+func TestOverridesExporterRing_scaleDown(t *testing.T) {
+	ringStore, closer := consul.NewInMemoryClient(ring.GetCodec(), log.NewNopLogger(), nil)
+	t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+	cfg1 := RingConfig{Enabled: true}
+	cfg1.KVStore.Mock = ringStore
+	cfg1.HeartbeatTimeout = 15 * time.Second
+
+	cfg1.InstanceID = "instance-1"
+	cfg1.InstanceAddr = "127.0.0.1"
+	i1, err := newRing(cfg1, log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	l1 := i1.lifecycler
+
+	cfg2 := cfg1
+	cfg2.InstanceID = "instance-2"
+	cfg2.InstanceAddr = "127.0.0.2"
+	i2, err := newRing(cfg2, log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	l2 := i2.lifecycler
+
+	// Register instances in the ring (manually, to be able to assign registered timestamps and tokens).
```

Comment:
[nit] Leaving this comment just in case you're to address the other comment too, otherwise we can skip this. It's really a minor thing.

```suggestion
	// Register instances in the ring (manually, to be able to assign tokens).
```